File: School/c++/lab3/wordLadder/wordLadder.cpp
#include "wordLadder.h"

void getDictionary(set<string>& dictionary)
{
    // declarations
    string fileName, dictionaryToken;
    ifstream fileOpener;

    // doing initialisation
    cout << "Dictionary file name? ";
    cin >> fileName;
    fileOpener.open(fileName);

    // if file not opened, prompt again
    while (!fileOpener.is_open()) {
        cout << "Unable to open that file. Try again. " << endl;
        cout << "Dictionary file name? ";
        cin >> fileName;
        fileOpener.open(fileName);
    }

    // insert words into dictionary set
    while (fileOpener >> dictionaryToken) {
        dictionary.insert(dictionaryToken);
    }

    fileOpener.close();
}

void getWord(set<string>& dictionary, string& word1, string& word2)
{
    // requesting words
    cout << endl << "Word #1 (or N to quit): ";
    cin >> word1;
    cout << "Word #2 (or N to quit): ";
    cin >> word2;

    // check several conditions. Reprompt if not met
    while (dictionary.count(word1) != 1 || dictionary.count(word2) != 1 || word1 == word2 || word1.size() != word2.size()) {
        if (word1 == "N" || word2 == "N") {
            break;
        }
        
        cout << "The two words must be found in the dictionary, different and have same length. " << endl;
        cout << endl << "Word #1 (or N to quit): ";
        cin >> word1;
        cout << "Word #2 (or N to quit): ";
        cin >> word2;
    }
}

stack<string> ladderSolver(set<string>& dictionary, string& word1, string& word2)
{
    // declarations
    set<string> usedWords; // set of used words
    stack<string> temp; // stack acts as first stack in qs
    queue<stack<string>> qs; // stack in queue to store all possible ladder
    stack<string> current; // act as main stack in calculation later

    temp.push(word1);
    qs.push(temp);
    
    while (!qs.empty()) {
        
        current = qs.front(); // take the front queue of stack as current stack to perform algorithm
        qs.pop(); // remove the stack already taken by current
        string currentWord = current.top(); // start with the topmost string of current stack

        for (int i = 0; i < currentWord.length(); i++) {
            for (int j = 97; j < 123; j++) {
                
                string newWord = currentWord;
                newWord[i] = static_cast<char>(j); // convert ASCII to alphabet character
                
                if (dictionary.count(newWord) == 1 && usedWords.count(newWord) != 1) {
                    
                    if (newWord == word2) {
                        current.push(newWord);
                        queue<stack<string>> empty; // create empty queue of stack
                        swap(qs, empty); // if succesful, make original qs empty to prevent going through further loop
                        break;
                    } else {
                        usedWords.insert(newWord);
                        stack<string> newStack = current;
                        newStack.push(newWord);
                        qs.push(newStack); // create new branch of possible solution as queue of stack
                    }
                }
            }
        }
    }

    return current;
}

void print(stack<string> S) // no reference used. So original stack will not affected. Safe to pop()
{
    while (!S.empty()) {
        cout << S.top() << " ";
        S.pop();
    }
    
    cout << endl;
}

File: School/JavaScript/wordLadder/script/WordLadder.js
// implement word ladder using breath first search
function wordLadder(word1, word2, dictionary)
{
    // declarations
    var usedWords = new Set(); // set of used words
    var temp = []; // stack acts as first stack in qs
    var qs = new Queue(); // stack in queue to store all possible ladder
    var current = []; // act as main stack in calculation later

    temp.push(word1);
    qs.enqueue(temp);
    dictionary[word1] = 1;
    var tempList = new Array(); // new array to store words changed to 1 in dictionary to restore them later
    tempList.push(word1);

    while (!qs.isEmpty()) {

        // take the front queue of stack as current stack to perform algorithm then remove the stack already taken by current
        current = qs.dequeue();
        var currentWord = current[current.length - 1]; // start with the topmost string of current stack

        for (var i = 0; i < currentWord.length; i++) {
            var alphabet = "abcdefghijklmnopqrstuvwxyz";

            for (var j = 0; j < alphabet.length; j++) {
                var newWord = currentWord.substring(0, i) + alphabet[j] + currentWord.substring(i + 1);

                if (dictionary.hasOwnProperty(newWord) && dictionary[newWord] != 1) {
                    if (newWord == word2) {
                        current.push(newWord);

                        while (!qs.isEmpty()) { // if successful, make original qs empty to prevent further loop
                            qs.dequeue();
                        }
                        break;

                    } else {
                        // perform required copying and formation of new stack for BFS
                        var newStack = current.slice(0); // deep copy
                        newStack.push(newWord);
                        qs.enqueue(newStack); // create new branch of possible solution as queue of stack
                        tempList.push(newWord);

                        // deals with dictionary and used words set
                        usedWords.add(newWord);
                        dictionary[newWord] = 1;
                    }
                }
            }
        }
    }

    var count = current.length;
    if (current[0] != word1 || current[count-1] != word2) {
        for (var i = 0; i < count; i++) {
            current.pop();
        }
    }

    // reset dictionary's value
    for (var i = 0; i < tempList.length; i++) {
        dictionary[tempList[i]] = 0;
    }
    tempList = [];
    
    return current;
}